### PR TITLE
P4-2115 diagnostics for production

### PR DESCRIPTION
**Pull request by martyn-w**
Remove PII (Personally identifiable information) from diagnostics for production and pre-production endpoints

### Jira link

P4-2115

### What?

I have added/removed/altered:

- [x] Allowed diagnostics for production, removing person and profile details from the output

### Why?

I am doing this because:

- to assist with supplier queries during/after go live

### Deployment risks (optional)

- minimal

### CODE REVIEW

**Comment by martyn-w** [comment]
Example output without PII

```
MOVE RECORD
-----------
id:	a8c06937-475b-4342-a5e2-5586cb28eecd
reference:	PXJ2479F
date:	2020-04-27
status:		requested
move type:	court_appearance
from location:	Oakhill Secure Training Centre
to location:	Houghton-le-Spring Youth Court
supplier:	
created at:	2020-04-27 12:46:06 +0100
updated at:	2020-05-05 12:16:39 +0100

MOVE EVENTS
-----------
(no events recorded)

JOURNEYS
--------
ID                                   	TIMESTAMP                  	STATE       	BILLABLE 	SUPPLIER 	FROM --> TO
855923bf-9cfd-4e53-a12c-1065d05ee954 	2020-04-27 05:00:00 +0100  	completed   	true     	Serco    	Oakhill Secure Training Centre --> Yarls Wood AIT
9b4b6c18-a6c0-454e-b9b4-c0534f14e885 	2020-04-27 05:43:00 +0100  	completed   	false    	Serco    	Yarls Wood AIT --> Weymouth Crown Court
bfbd65c4-8cbc-4918-8ce1-639af8a4ff38 	2020-04-28 05:43:00 +0100  	completed   	false    	Serco    	Weymouth Crown Court --> Yarls Wood AIT
4cc4e9a6-3f30-4fbc-b267-d5a32f4401ea 	2020-04-28 06:47:00 +0100  	completed   	true     	Serco    	Yarls Wood AIT --> Houghton-le-Spring Youth Court

JOURNEY EVENTS
--------------
Oakhill Secure Training Centre --> Yarls Wood AIT (855923bf-9cfd-4e53-a12c-1065d05ee954)
  EVENT          	TIMESTAMP                  	PARAMS
  lockout        	2020-04-27 05:43:00 +0100  	-

Yarls Wood AIT --> Weymouth Crown Court (9b4b6c18-a6c0-454e-b9b4-c0534f14e885)
  EVENT          	TIMESTAMP                  	PARAMS
  redirect       	2020-04-27 05:43:00 +0100  	-

Weymouth Crown Court --> Yarls Wood AIT (bfbd65c4-8cbc-4918-8ce1-639af8a4ff38)
  EVENT          	TIMESTAMP                  	PARAMS
  redirect       	2020-04-28 05:43:00 +0100  	-

Yarls Wood AIT --> Houghton-le-Spring Youth Court (4cc4e9a6-3f30-4fbc-b267-d5a32f4401ea)
  (no events recorded)


WEBHOOK NOTIFICATIONS
---------------------
TYPE              	DELIVERED AT               	ATTEMPTS 	ENDPOINT
create_move       	2020-04-27 12:46:06 +0100  	1        	http://example.test

EMAIL NOTIFICATIONS
-------------------
TYPE              	DELIVERED AT               	ATTEMPTS 	EMAIL
create_move       	2020-04-27 12:46:06 +0100  	1        	foo@example.test

```

File: app/controllers/diagnostics_controller.rb [hunk 1]
```diff
@@ -1,12 +1,17 @@
 # frozen_string_literal: true
 
-class DiagnosticsController < ApplicationController
-  before_action :doorkeeper_authorize!, unless: -> { Rails.env.development? }
+class DiagnosticsController < ApiController
```
Comment (willfish) [inline app/controllers/diagnostics_controller.rb:3]: Wasn't sure whether we're getting anything from the ApiController that we don't want?

Comment (willfish) [reply app/controllers/diagnostics_controller.rb:3]: I guess if it works it works 👍

File: app/controllers/diagnostics_controller.rb [hunk 2]
```diff
@@ -1,12 +1,17 @@
...
+  def set_content_type
+    self.content_type = 'text/plain'
+  end
+
+  def move
+    move = Move.accessible_by(current_ability).find_by(id: params[:id]) || Move.accessible_by(current_ability).find_by(reference: params[:id])
```
Comment (willfish) [inline app/controllers/diagnostics_controller.rb:9]: Safer!

File: app/services/diagnostics/move_inspector.rb [hunk 3]
```diff
@@ -19,24 +20,26 @@ def generate
       @output << "time due:\t#{move.time_due}\n" if move.time_due.present?
       @output << "status:\t\t#{move.status}\n"
       @output << "cancel reason:\t#{move.cancellation_reason}\n" if move.cancellation_reason.present?
-      @output << "cancel comment:\t#{move.cancellation_reason_comment}\n" if move.cancellation_reason_comment.present?
-      @output << <<~ENDDETAILS
+      @output << "cancel comment:\t#{move.cancellation_reason_comment}\n" if include_person_details && move.cancellation_reason_comment.present?
+      @output << <<~ENDDETAILS1
```
Comment (willfish) [inline app/services/diagnostics/move_inspector.rb:24]: I think the heredoc naming doesn't actually get memoized beyond the scope of opening/closing the heredoc and I'm not sure that 1 and 2 add much information to other developers reading this.

File: config/routes.rb [hunk 4]
```diff
@@ -10,11 +10,7 @@
 
   get '/health', to: 'status#health', format: :json
   get '/ping', to: 'status#ping', format: :json
-
-  # diagnostics should only be available on localhost, dev, staging and uat; not on pre-prod or production
-  if Rails.env.development? || ENV.fetch('HOSTNAME', 'UNKNOWN') =~ /(\-(dev|staging|uat)\-)/i
-    get '/diagnostics/moves/:id', to: 'diagnostics#moves'
```
Comment (willfish) [inline config/routes.rb:16]: Singular is definitely clearer

**Review by willfish — APPROVED** [review]

File: app/controllers/diagnostics_controller.rb (hunk 1 above)
Comment (martyn-w) [reply app/controllers/diagnostics_controller.rb:3]: the API controller brings in authorisation; the only annoyance is it defaults to json content type which has to be overridden in this controller.

File: app/services/diagnostics/move_inspector.rb [hunk 5]
```diff
@@ -51,6 +54,7 @@ def generate
       if move.journeys.any?
         @output << "#{'ID'.ljust(37)}\t#{'TIMESTAMP'.ljust(27)}\t#{'STATE'.ljust(12)}\t#{'BILLABLE'.ljust(9)}\t#{'SUPPLIER'.ljust(9)}\tFROM --> TO\n"
         move.journeys.default_order.each do |journey| # NB use each to preserve sort order
+          # NB only show event details if include_person_details==true, as they could contain personal details
```
Comment (smoothcontract) [inline app/services/diagnostics/move_inspector.rb:57]: I can't see the code that this comment relates to - this block doesn't seem to be checking the value of `include_person_details`.

File: app/services/diagnostics/move_inspector.rb [hunk 6]
```diff
@@ -107,63 +112,65 @@ def generate
...
+              @output << "nomis_alert_description:\t#{answer.nomis_alert_description}\n"
+              @output << "imported_from_nomis:\t#{answer.imported_from_nomis}\n"
+              @output << "---\n"
+            end
+          else
+            @output << "(no assessment answers recorded)\n"
           end
         else
-          @output << "(no assessment answers recorded)\n"
+          @output << "(no profile associated with move)\n"
```
Comment (smoothcontract) [inline app/services/diagnostics/move_inspector.rb:172]: This is a very long method - not the end of the world as it's all pretty simple stuff but perhaps potentially a candidate for splitting into smaller sections?

File: config/routes.rb [hunk 7]
```diff
@@ -10,11 +10,7 @@
...
-  end
```
Comment (smoothcontract) [inline config/routes.rb:17]: 👍 Much cleaner to use the existing auth support

File: lib/tasks/inspect_move.rake [hunk 8]
```diff
@@ -7,6 +7,7 @@ namespace :inspect do
     move = Move.find_by(id: args[:id_or_ref]) || Move.find_by(reference: args[:id_or_ref])
     abort "Could not find move record with id or reference: #{args[:id_or_ref]}" if move.blank?
 
-    puts Diagnostics::MoveInspector.new(move).generate
+    # OK to return personal details if running as a rake task on ther server
```
Comment (smoothcontract) [inline lib/tasks/inspect_move.rake:10]: typo - `ther`

File: app/services/diagnostics/move_inspector.rb (hunk 3 above)
Comment (martyn-w) [reply app/services/diagnostics/move_inspector.rb]: now renamed with new and improved names

**Review by smoothcontract — COMMENTED** [review]
Just some minor comments, looks very useful!

File: app/services/diagnostics/move_inspector.rb (hunk 5 above)
Comment (martyn-w) [reply app/services/diagnostics/move_inspector.rb:57]: oops, the comment was meant for the previous block, now fixed

File: lib/tasks/inspect_move.rake (hunk 8 above)
Comment (martyn-w) [reply lib/tasks/inspect_move.rake:10]: fixed

File: app/services/diagnostics/move_inspector.rb (hunk 6 above)
Comment (martyn-w) [reply app/services/diagnostics/move_inspector.rb:172]: yes.. fair enough - will save that for a rainy day

**Review by willfish — APPROVED** [review]

**Pull request merged**